tasks/read-a-configuration-file/src/main.rs

```diff
@@ -0,0 +1,158 @@
+use std::path::Path;
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq)]
+enum ConfigType {
+    String,
+    Boolean,
+    Vector,
+}
+
+#[derive(Debug)]
+struct ConfigVariable {
+    var_type: ConfigType,
+    value: String,
+}
+
+impl ConfigVariable {
+    fn extract_value_as_string(&self, typ: ConfigType) -> Result<String, String> {
+        if self.var_type == typ {
+            Ok(self.value.to_owned())
+        } else {
+            Err(format!("Incorrect variable type. Expected: {:?}. Actual: {:?}",
+                        typ,
+                        self.var_type))
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ConfigParams {
+    params: HashMap<String, ConfigVariable>,
+}
+
+
+impl ConfigParams {
+    fn new() -> ConfigParams {
+        ConfigParams { params: HashMap::new() }
+    }
+
+    fn parse<P: AsRef<Path>>(path: P) -> ConfigParams {
+        let conf_file = File::open(path).unwrap();
+        let content = BufReader::new(conf_file);
+
+        let is_not_comment = |x: &Result<String, io::Error>| {
+            match x {
+                &Err(_) => false,
+                &Ok(ref line) => {
+                    !(line.starts_with("#") || line.starts_with(";") || line.is_empty())
+                }
+            }
+
+        };
+        let mut params = ConfigParams::new();
+        for line in content.lines().filter(is_not_comment) {
+            if line.is_ok() {
+                params.update_config(line.unwrap());
+            }
+
+        }
+
+        params
+    }
+    // Will parse the line and update the internal structure
+    fn update_config(&mut self, line: String) {
+        // First space splits the key from value, then if a list is separated by a comma
+        let mut parts = line.splitn(2, " ").map(|x| x.to_owned());
+        let key = parts.next().unwrap().to_lowercase();
+        match parts.next() {
+            None => {
+                self.params.insert(key,
+                                   ConfigVariable {
+                                       var_type: ConfigType::Boolean,
+                                       value: "true".to_owned(),
+                                   });
+            }
+            Some(value) => {
+                if value.contains(",") {
+                    self.params.insert(key,
+                                       ConfigVariable {
+                                           var_type: ConfigType::Vector,
+                                           value: value,
+                                       });
+                } else {
+                    self.params.insert(key,
+                                       ConfigVariable {
+                                           var_type: ConfigType::String,
+                                           value: value,
+                                       });
+                }
+            }
+        }
+
+    }
+
+    fn string_param(&self, key: &str) -> Result<String, String> {
```

Comment:
I wonder if you could make these a generic like `param::<String>::("fullname")` instead... Might be interesting to try if you'd like! :)